content/react/en/creating-addons.md

```diff
@@ -0,0 +1,429 @@
+import React from 'react';
+
+import { addons, types } from '@storybook/addons';
+import { AddonPanel } from '@storybook/components';
+import { ADDON_ID, PANEL_ID } from './constants';
+import { Content } from './panel';
+
+addons.register(ADDON_ID, () => {
+  addons.add(PANEL_ID, {
+    title: 'design assets',
+    type: types.PANEL,
+    render: ({ active, key }) => (
+      <AddonPanel active={active} key={key}>
+        <Content />
+      </AddonPanel>
+    ),
+  });
+});
+```
+
+```js
+// panel.js
+import React, { Fragment, useMemo, ReactElement } from 'react';
+
+import { useParameter, useAddonState, useStorybookState } from '@storybook/api';
+import { styled } from '@storybook/theming';
+import { ActionBar } from '@storybook/components';
+import { PARAM_KEY, ADDON_ID } from './constants';
+
+interface AssetDescription {
+  url: string;
+  name: string;
+}
+
+type Results = (string | AssetDescription)[];
+type Selected = number;
+
+const Iframe = styled.iframe({
+  width: '100%',
+  height: '100%',
+  border: '0 none',
+});
+const Img = styled.img({
+  width: '100%',
+  height: '100%',
+  border: '0 none',
+  objectFit: 'contain',
+});
+
+const Asset = ({ url }: { url: string | undefined }): ReactElement => {
+  if (!url) {
+    return null;
+  }
+  if (url.match(/\.(png|gif|jpeg|tiff|svg|anpg|webp)/)) {
+    // do image viewer
+    return <Img alt="" src={url} />;
+  }
+  if (url.match(/\.(mp4|ogv|webm)/)) {
+    // do video viewer
+    return <div>not implemented yet, sorry</div>;
+  }
+
+  return <Iframe title={url} src={url} />;
+};
+
+const getUrl = (input: AssetDescription | string): string => {
+  return typeof input === 'string' ? input : input.url;
+};
+
+export const Content = () => {
+  const results = useParameter<Results>(PARAM_KEY, []);
+  const [selected, setSelected] = useAddonState<Selected>(ADDON_ID, 0);
+  const { storyId } = useStorybookState();
+
+  if (results.length === 0) {
+    return null;
+  }
+
+  if (results.length && !results[selected]) {
+    setSelected(0);
+    return null;
+  }
+
+  return (
+    <Fragment>
+      <Asset url={url} />
+      {results.length > 1 ? (
+        <ActionBar
+          actionItems={results.map((i, index) => ({
+            title: typeof i === 'string' ? `asset #${index + 1}` : i.name,
+            onClick: () => setSelected(index),
+          }))}
+        />
+      ) : null}
+    </Fragment>
+  );
+};
+```
+
+## Extracting your addon into it's own package
+
+After your addon is working as intended, you will likely want to extract it into it's own package.
+
+If it's useful to you, it's likely useful to someone else as well.
+
+You can open a PR on the storybook documentation to have your addon featured there.
+
+## What's next?
+
+This is an example of an addon that uses parameters and displays it in a panel, but that's just one of the many options you have.
+
+With these principles you're able to display you custom UI in a variety of places in the storybook UI.
+
+You can:
```

Comment:
Could you please permalink each item to a code example from one of our addons to make it more convenient to find these techniques.

Reply:
✅